### ISSUE TITLE
Add cylinder as optional entity collider

### ISSUE THREAD

**Issue by MarkuBu** [minetest/minetest#6812]
Unity uses capsules as collider object

**Comment by paramat** [comment]
MT collisionboxes don't rotate with the object, they stay axis aligned just like selectionboxes, so the yaw of the mob doesn't matter.
However a cylinder does have some advantage, but i think a rotating rectangle or 2 axis-aligned rectangles bring much more advantage for probably similar processing. A cylinder doesn't address objects that are long or wide, and we now have the ability for players to be any shape.
A cylinder won't help for vehicles that are long or wide.
I'd rather see this work go towards something that copes with long or wide objects.

**Comment by MarkuBu** [comment]
Of course this should be optional. And as I mentioned this would be most interesting for human like entities.

But it has some advantages over a box. If the box doesn't rotate, an entity slides along a node until the edge of the box is reached. The mesh rotates to the new direction, but still moves straight. This can look odd and can nearly stop an entity. Also entities can "intertwine" if they try to pass each other

If you rotate the collision box, mobs will stuck wherever the space is smaller than the diagonal size of the collision box. This would make things much worse than it is now

Even if Minetest is a "box" game a cylindrical collider would improve the gameplay because boxes are "jerky"

**Comment by MarkuBu** [comment]
Performance is not always a reason to reject new features. I know that mobile and low end devices are important for Minetest, but there are a lot of people with enough power to use these features easily. 

I know that you are concerned that some people can't play on some servers or subgames and I can understand that, but always reject new features only because of some more CPU usage will make Minetest obsolete one day because each old PC or smartphone will die one day and will be replaced by some new and more powerful device.

So please don't reject each suggestion that add's a little bit of CPU time

**Comment by Ekdohibs** [comment]
Note that we have existing problems with collision already (for instance, it is sometimes possible to enter a position but not go back), that this could fix. So, especially if this fixes those kind of bugs, I think it would be good to have it.

**Comment by MarkuBu** [comment]
Yes, there a lot of issues with collision. I reported some of them myself
#6807
#6749

I would love to help fixing them, but my C++ skills are not enough to make larger changes and I don't really know enough about the engine. And next year I have much less time for such things.

**Comment by stujones11** [comment]
> a cylinder will be much more intensive to process collisions for

Spherical/cylindrical collisions need not be any computationally expensive than aabb boxes if it's done correctly, however, I agree that the existing collision system need fixing before we even consider other shapes.

**Comment by rubenwardy** [comment]
Cylinder collisions will be less intensive as you only need to check radius and then upper/lower bounds. Assuming no rotation. Not sure about cylinder/cube collisions

**Comment by stujones11** [comment]
> Not sure about cylinder/cube collisions

Very easy for axis axis aligned boxes, you simply check if they fall in the same height range and only check circle intersects rectangle if they do :)

**Comment by azekillDIABLO** [comment]
I'm a bit against this for two reasons:
* I think this will add a lot more collision to calculate, and will probably bring lag and bugs
* I have no idea how you think to render it with the outline of collision box

A rotating collision box would be better.

**Comment by benrob0329** [comment]
>I think this will add a lot more collision to calculate, and will probably bring lag and bugs

>Cylinder collisions will be less intensive as you only need to check radius and then upper/lower bounds.

>I have no idea how you think to render it with the outline of collision box

Like you would any other outline.

**Comment by MarkuBu** [comment]
> I think this will add a lot more collision to calculate

What?

> I have no idea how you think to render it with the outline of collision box

Is that necessary at all? I have disabled them because it isn't. Have you ever seen someone playing Minecraft with enabled collision boxes? I've seen a lot of videos and no one plays with visual collision boxes. It is nonsense and makes the visual appearance worse. That's one of the default settings I've never understood.

And by the way: the collision box is never visual. What you see is the selection box

> A rotating collision box would be better.

Not in any situation. It makes sense e.g. for a car or maybe a cow (long, but narrow), but my suggestion is only for human like mobs and you would make them unusable with a rotating box

**Comment by stujones11** [comment]
> And by the way: the collision box is never visual. What you see is the selection box

This is a good point, ray-casting currently uses the selection-box, anyway. I see no reason to change that, collision can be different to selection for numerous reasons.

**Comment by MarkuBu** [comment]
Maybe interesting to this topic is this video
https://www.youtube.com/watch?v=FtNcFw3GD6A

It shows all hitboxes of (nearly) all Minecraft Mobs. Minecaft uses the same technique as Minetest, square, non rotating boxes, except the ender dragon (look at 14:20) which uses a multi hitbox model

I don't know how much more a cylinder would cost to calculate. I assume you need sine and cosine. As alternative an approximate shape like a hexagon or octagon would work too.

The reason for my suggestion is to get rid of the edges of the boxes. They cause problems in certain situations and even a hexagon would make movement of a mob much easier.

**Comment by stujones11** [comment]
I don't know how much more a cylinder would cost to calculate.

No complex mathematics are necessary whatsoever, IIRC you just compare the squared rads to avoid using sqrt.

**Comment by paramat** [comment]
>  but always reject new features only because of some more CPU usage will make Minetest obsolete one day because each old PC or smartphone will die one day and will be replaced by some new and more powerful device.
So please don't reject each suggestion that add's a little bit of CPU time

This is a misrepresentation of my arguments. Collision processing is performance critical, and cylinder collisions are probably more complex than boxes to process. I'm balancing that against a very minor gain.

> Note that we have existing problems with collision already (for instance, it is sometimes possible to enter a position but not go back), that this could fix.

I don't know yet of any this could fix, it won't affect #6807 or the 'slab trap' issue because a cylinder acts identically in the y direction.

Quoting rubenwardy:

> Cylinder collisions will be less intensive as you only need to check radius and then upper/lower bounds. Assuming no rotation. Not sure about cylinder/cube collisions

Cylinder-box collisions will be the vast majority of collisions, so cylinder-cylinder being simpler is not a valid argument.

First we need profiling to see exactly how intensive cylinder-box collisions are, if it turns out no more intensive than box-box i'll be happy to support. But my geometric intuition is saying that it will be more intensive, something to investigate.

The fact that MC uses boxes may suggest that is optimal and doesn't cause significant issues.

**Comment by stujones11** [comment]
> But my geometric intuition is saying that it will be more intensive, something to investigate.

For non-axis aligned collisions you are likely correct, however, there are some rather neat little tricks you can use when the alignment in known. Definitely worth investigating, IMO.

**Comment by appgurueu** [comment]
I would expect cylindrical collisions to be slightly less efficient and nastier than AABB collisions [^1], but worrying about that is pure premature optimization, especially as an AABB can always be used to filter out the vast majority of collisions that don't happen before applying the more granular cylindrical collider.

It seems to me that cylinders are a reasonably simple and useful collider shape for entities that mostly move and collide in the XZ plane and rotate around Y, especially as (except for the top and bottom) they have no edges to get stuck on, unlike boxes.

I would not consider this low priority.

[^1]: Cylinder - cylinder collisions will require solving quadratic equations for the point in time at which the circles collide. Cylinder - box collisions are also a bit nasty. Raycasts also end up with a quadratic equation.
